spec/acts_as_tag_pgarray/taggable_spec.rb

```diff
@@ -1,162 +1,188 @@
-    end
-    it "defines named scope not to match all tags" do
-      expect(User).to respond_to(:without_all_colors)
-    end
-  end
-
-  describe "#taggable_array_on" do
-    it "defines named scope to match any tags" do
-      expect(User).to respond_to(:with_any_codes)
-    end
-    it "defines named scope to match all tags" do
-      expect(User).to respond_to(:with_all_codes)
-    end
-    it "defines named scope not to match any tags" do
-      expect(User).to respond_to(:without_any_codes)
-    end
-    it "defines named scope not to match all tags" do
-      expect(User).to respond_to(:without_all_codes)
-    end
-  end
-
-  it "should define table name un-ambiguously" do
-    sql = User.with_any_sizes(["small"]).to_sql
-    expect(sql).to eql("SELECT \"users\".* FROM \"users\" WHERE (users.sizes && ARRAY['small']::text[])")
-
-    sql = User.with_all_sizes(["small"]).to_sql
-    expect(sql).to eql("SELECT \"users\".* FROM \"users\" WHERE (users.sizes @> ARRAY['small']::text[])")
-
-    sql = User.without_any_sizes(["small"]).to_sql
-    expect(sql).to eql("SELECT \"users\".* FROM \"users\" WHERE NOT (users.sizes && ARRAY['small']::text[])")
-
-    sql = User.without_all_sizes(["small"]).to_sql
-    expect(sql).to eql("SELECT \"users\".* FROM \"users\" WHERE NOT (users.sizes @> ARRAY['small']::text[])")
-  end
-
-  it "should work with ::text typed array" do
-    expect(User.with_any_sizes(["small"])).to match_array([@user2, @user3, @admin2])
-    expect(User.with_all_sizes(["small", "large"])).to match_array([@user2, @user3, @admin2])
-    expect(User.without_any_sizes("medium")).to match_array([@user2, @admin2])
-    expect(User.without_all_sizes("medium")).to match_array([@user2, @admin2])
-  end
-
-  it "should work with ::integer typed array" do
-    expect(User.with_any_codes([123])).to match_array([@user2, @user3, @admin1, @admin2])
-    expect(User.with_all_codes([123, 789])).to match_array([@user2, @user3, @admin1])
-    expect(User.without_any_codes(456)).to match_array([@user2])
-    expect(User.without_all_codes(456)).to match_array([@user2])
-  end
-
-  describe "#with_any_tags" do
-    it "returns users having any tags of args" do
-      expect(User.with_any_colors(["red", "blue"])).to match_array([@user1, @user2, @user3, @admin2])
-      expect(User.with_any_colors("red, blue")).to match_array([@user1, @user2, @user3, @admin2])
-    end
-  end
-
-  describe "#with_all_tags" do
-    it "returns users having all tags of args" do
-      expect(User.with_all_colors(["red", "blue"])).to match_array([@user1])
-      expect(User.with_all_colors("red, blue")).to match_array([@user1])
-    end
-  end
-
-  describe "#without_any_tags" do
-    it "returns users not having any tags of args" do
-      expect(User.without_any_colors(["red", "blue"])).to match_array([@admin1])
-      expect(User.without_any_colors("red, blue")).to match_array([@admin1])
-    end
-  end
+  context "without allow_list" do
```

Comment:
Top level context has become:
- "without allow_list"
     - (all other cases)
- "with allow_list"
   - (all other cases)
   
I don't feel that "allow_list" is such an important feature to warant top ranking in testsuit hierarchy. Can we append to previous hierarchy? Example

- (all other cases)
- "#allow_list#
  - without allow_list
    - (tests)  
  - with allow_list
    - (test)

Reply:
what do you think to something like:
* without any configurations
* with configurations
  * #allow_list

I put all the other cases inside a context to be sure that the setup for those other cases does not pollute the test scenarios for allow list and vice versa. If I just remove the context and let all other cases direct in the top level, it will run a lot of setup code on before for the allow list tests that are not necessary, won't be useful and can generate unintended results. Even thinking about test performance, adding more options, and having this top-level cases running every time is not good.

I considered writing the validations tests(allow list now, and maybe a block list, quantity limit, and more in the future) in a different file, so we would have all normal and top case scenarios here in this file, and, for now, allow list tests in another file. What do you think of this approach?